File: project/adam/C/utils/maxent_utils.cpp
/*********************************************************************
Copyright (C) 2009 Hewlett-Packard Development Company, L.P.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*********************************************************************/

#include <string.h>
#include "maxent_utils.h"
#include <default_list.h>
#include "token.h"
#include "token_feature.h"
#include "repr.h"
#include <limits.h>
#include "hash.h"
#include <math.h>
#include "sentence.h"

unsigned long create_context(default_list feature_type_list, int left_window, int right_window, int i, me_context_type& context) {
    int j, k, n;
    string hash_string;
    n = default_list_length(feature_type_list);
    context.clear();
    for (j=-left_window+1; j<right_window; j++) {
        if (i+j<0 || i+j>=n) {
            //printf("%d ",i+j);
        } else {
            token_feature *ft = (token_feature *)default_list_get(feature_type_list, i+j);
            char key[1024];
            char rstr[1024];
            
            for (k = 0; k < FT_CHAR_MAP_LEN; k++) {
                sprintf(key,"char_%d[%03d]",j,k);
                context.push_back(make_pair(string(key), (float)ft->char_vector[k]));
                sprintf(key,"char_%d[%03d]=%d ",j,k,ft->char_vector[k]);
                hash_string.append(key);
            }
            if (ft->word==TRUE) {
                sprintf(key,"word_%d",j);
                context.push_back(make_pair(string(key), 1.0));
                sprintf(key,"word_%d=1 ",j);
                hash_string.append(key);
            }
            if (ft->number==TRUE) {
                sprintf(key,"number_%d",j);
                context.push_back(make_pair(string(key), 1.0));
                sprintf(key,"number_%d=1 ",j);
                hash_string.append(key);
            }
            //char key[1024];
            //char rstr[1024];
            //repr_string(rstr,ft->string);
            //sprintf(key,"token_%d='%s'",j,rstr);
            //context.push_back(make_pair(string(key), 1.0));
            
            //sprintf(key,"word_%d",j);
            //if (ft->word==TRUE) {
            //    context.push_back(make_pair(string(key), 1.0));
            //    sprintf(key,"stem_%d='%s'",j,ft->stemmed);
            //    context.push_back(make_pair(string(key), 1.0));
            //    sprintf(key,"capped_%d",j);
            //    if (ft->capped==TRUE) {
            //        context.push_back(make_pair(string(key), 1.0));
            //    } else {
            //        context.push_back(make_pair(string(key), 0.0));
            //    }
            //    sprintf(key,"upper_%d",j);
            //    if (ft->upper==TRUE) {
            //        context.push_back(make_pair(string(key), 1.0));
            //    } else {
            //        context.push_back(make_pair(string(key), 0.0));
            //    }
            //    sprintf(key,"number_%d",j);
            //    if (ft->number==TRUE) {
            //        context.push_back(make_pair(string(key), 1.0));
            //    } else {
            //        context.push_back(make_pair(string(key), 0.0));
            //    }
            //    sprintf(key,"incnum_%d",j);
            //    if (ft->incnum==TRUE) {
            //        context.push_back(make_pair(string(key), 1.0));
            //    } else {
            //        context.push_back(make_pair(string(key), 0.0));
            //    }
            //} else {
            //    context.push_back(make_pair(string(key), 0.0));
            //    //sprintf(key,"stem_%d=''",j);
            //    //context.push_back(make_pair(string(key), 1.0));
            //    for (k = 1; k < FT_CHAR_MAP_LEN; k++) {
            //        sprintf(key,"char_%d='_%03d_'",j,k);
            //        context.push_back(make_pair(string(key), (float)ft->char_vector[k]));
            //    }
            //}
        }
    }
    // cout << hash_string << "\n" << sdbm_string(hash_string) << endl;
    return sdbm_string(hash_string);
}

void create_model(MaxentModel& m, default_list feature_type_list, default_list label_list, int left_window, int right_window) {
    int i, j, k, n;
    me_context_type context;
    me_outcome_type outcome;

    sv_vector vect_E = sv_new(ULONG_MAX);
    sv_vector vect_I = sv_new(ULONG_MAX);

    n = default_list_length(feature_type_list);
    for (i = 0; i<n; i++) {
        char *c = (char *)default_list_get(label_list, i);
        token_feature *tf = (token_feature *)default_list_get(feature_type_list,i);
        if (tf->word != TRUE && tf->char_vector[0] != tf->length) {
            unsigned long index = create_context(feature_type_list,left_window,right_window,i,context);
            if (strcmp("E",c) == 0) {
                double v = sv_get_element_value(vect_E,index);
                sv_set_element(vect_E,index,v+1.0);
            } else {
                double v = sv_get_element_value(vect_I,index);
                sv_set_element(vect_I,index,v+1.0);
            }
        }
    }

    for (i = 0; i<n; i++) {
        char *c = (char *)default_list_get(label_list, i);
        token_feature *tf = (token_feature *)default_list_get(feature_type_list,i);
        if (tf->word != TRUE && tf->char_vector[0] != tf->length) {
            unsigned long index = create_context(feature_type_list,left_window,right_window,i,context);
            if (strcmp("E",c) == 0) {
                double v = sv_get_element_value(vect_E,index);
                if (v > 0) {
                    m.add_event(context,c,v);
                }
                sv_set_element(vect_E,index,0);
            } else {
                double v = sv_get_element_value(vect_I,index);
                if (v > 0) {
                    m.add_event(context,c,v);
                }
                sv_set_element(vect_I,index,0);
            }
        }
    }
}

void label_sentences(MaxentModel& m, default_list feature_type_list, default_list label_list, int left_window, int right_window) {
    int i,n;
    char E[2] = "E";
    char I[2] = "I";
    me_context_type context;
    
    n = default_list_length(feature_type_list);
    for (i = 0; i<n; i++) {
        token_feature *tf = (token_feature *)default_list_get(feature_type_list,i);
        float p = 0.0;
        
        if (tf->word != TRUE && tf->char_vector[0] != tf->length) {
            create_context(feature_type_list,left_window,right_window,i,context);
            p = m.eval(context,"E");
            if (p>=0.4) {
                default_list_append(label_list,E);
            } else {
                default_list_append(label_list,I);
            }
        } else {
            default_list_append(label_list,I);
        }
    }
}

int create_sentences(MaxentModel& m, default_list sentence_list, char *buffer, default_list feature_type_list, default_list label_list, char *filename, char *licensename, int id) {
    token_feature *ft = NULL;
    char *t = NULL;
    sv_vector vect;
    sentence *st = NULL;
    int i;

    vect = sv_new(ULONG_MAX);
    ft = (token_feature *)default_list_get(feature_type_list,0);
    int start = ft->start;
    for (i = 0; i<default_list_length(feature_type_list); i++) {
        ft = (token_feature *)default_list_get(feature_type_list,i);
        t = (char *)default_list_get(label_list,i);
        if (ft->word == TRUE) {
            double v = 0;
            unsigned long int index = 0;
            index = sdbm(ft->stemmed);
            v = sv_get_element_value(vect,index);
            sv_set_element(vect,index,v+1.0);
        }
        if (strcmp(t, "E")==0 || i == default_list_length(feature_type_list)-1) {
            if (i < default_list_length(feature_type_list)-1 && sv_nonzeros(vect)<2) {

            } else {
                double norm = sqrt(sv_inner(vect,vect));
                if (norm == 0) {
                    continue;
                }
                vect = sv_scalar_mult(vect,1.0/norm);
                st = sentence_create(buffer,start,ft->end,i,filename,licensename,id,vect);

                default_list_append(sentence_list,st);

                vect = sv_new(ULONG_MAX);
                start = ft->end;
            }
        }
    }
    return 0;
}

File: project/adam/python/utils/maxent_utils.py
#!/usr/bin/python

## 
## Copyright (C) 2007 Hewlett-Packard Development Company, L.P.
## 
## This program is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License
## version 2 as published by the Free Software Foundation.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License along
## with this program; if not, write to the Free Software Foundation, Inc.,
## 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
##

import parser

import re, unicodedata
from maxent import MaxentModel
from Stemmer import Stemmer
import math
import stopwords

def get_context(features,index,left,right):
    n = len(features)
    a = max(0,index-left)
    b = min(n,index+right+1)
    window = left+right+1
    d = features[a:b]
    if len(d)<window:
        if a == 0:
            p = 0
        else:
            p = window
        while len(d)<window:
            d.insert(p,{'token':'__BLANK__'})
    context = []
    for i in xrange(window):
        for k in d[i]:
            context.append('%s_%d=%r' % (k,i-left,d[i][k]))
    return context

def sentences(features,model,left,right):
    sents = []
    n = len(features)
    for i in xrange(n):
        if not features[i].get('word',True) and not features[i].get('white',True):
            E = model.eval(get_context(features,i,left,right),'E')
            if E>0.5:
                sents.append(i)
    return sents

def train_sentences(files,model,left,right):
    file_features = {}
    file_sentences = {}
    for file in files:
        text = open(file).read()

        sentences = parser.RE_SENTENCE.findall(text)
        for i in range(1, len(sentences)):
            extra = parser.RE_START_NONWORD.match(sentences[i])
            if extra:
                sentences[i-1] = "%s%s" % (sentences[i-1],sentences[i][extra.start():extra.end()])
                sentences[i] = sentences[i][extra.end():]
        if len(sentences)>0:
            file_features[file] = []
            file_sentences[file] = []
        for s in sentences:
            f = parser.features(''.join(s))
            file_sentences[file].append((len(f)-1)+(len(file_features[file])))
            file_features[file].extend(f)

    context_dict = {}

    for file in files:
        n = len(file_features[file])
        for i in xrange(n):
            if not file_features[file][i].get('word',True) and not file_features[file][i].get('white',True):
                context = get_context(file_features[file],i,left,right)
                label = 'I'
                if i in file_sentences[file]:
                    label = 'E'
                key = ' '.join(context)+' '+label
                context_dict[key] = context_dict.get(key,[context,label,0])
                context_dict[key][2] += 1
                if context_dict.get(' '.join(context)+' I',False) and context_dict.get(' '.join(context)+' E',False):
                    print "Conflicting features in file %s with context:\n\n%s" % (file,context)


    model.begin_add_event()
    for k in context_dict:
        model.add_event(context_dict[k][0],context_dict[k][1],context_dict[k][2])
    model.end_add_event()
    model.train(1000, 'lbfgs')

    correct = 0
    for k in context_dict:
        E = model.eval(context_dict[k][0],'E')
        if (E>0.5 and context_dict[k][1] == 'E') or (E<0.5 and context_dict[k][1] == 'I'):
            correct += 1
        else:
            print "Missed: %s" % k

    print "Classification was %d/%d." % (correct, len(context_dict))

    return model

def sentence_word_arrays(features,sents):
    pass

def sentence_stem_arrays(features,sents):
    n = len(features)
    stems = [[]]
    for i in xrange(n):
        t = features[i]['stem']
        if len(t)>0:
            stems[-1].append(t)
        if i in sents:
            stems.append([])

    return stems

def sentence_byte_offsets(features,sents):
    n = len(features)
    byte_offsets = [[0,0]]
    for i in xrange(n):
        byte_offsets[-1][1] += features[i]['length']
        if i in sents:
            byte_offsets.append([byte_offsets[-1][1],byte_offsets[-1][1]])

    return byte_offsets

